In [57]:
import pandas as pd
import numpy as np
from scipy import stats
from sklearn.neighbors import KNeighborsRegressor

In [59]:
df = pd.read_csv("Fraud Detection Dataset.csv")

In [61]:
df.head()

Transaction_ID  User_ID  Transaction_Amount Transaction_Type  \
0             T1     4174             1292.76   ATM Withdrawal   
1             T2     4507             1554.58   ATM Withdrawal   
2             T3     1860             2395.02   ATM Withdrawal   
3             T4     2294              100.10     Bill Payment   
4             T5     2130             1490.50      POS Payment   

   Time_of_Transaction Device_Used       Location  \
0                 16.0      Tablet  San Francisco   
1                 13.0      Mobile       New York   
2                  NaN      Mobile            NaN   
3                 15.0     Desktop        Chicago   
4                 19.0      Mobile  San Francisco   

   Previous_Fraudulent_Transactions  Account_Age  \
0                                 0          119   
1                                 4           79   
2                                 3          115   
3                                 4            3   
4                                 2           57   

   Number_of_Transactions_Last_24H Payment_Method  Fraudulent  
0                               13     Debit Card           0  
1                                3    Credit Card           0  
2                                9            NaN           0  
3                                4            UPI           0  
4                                7    Credit Card           0

### Hypothesis
Null hypotheis: there are no features that distinguish between fraudulent transactions and non-fraudulent transactions<br>
Alternate Hypothesis: there is at least one feature that distinguishes between fraudulent transactions and non-fraudulent transactions

### Model Selection
Classification model - kNN

In [65]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51000 entries, 0 to 50999
Data columns (total 12 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Transaction_ID                    51000 non-null  object 
 1   User_ID                           51000 non-null  int64  
 2   Transaction_Amount                48480 non-null  float64
 3   Transaction_Type                  51000 non-null  object 
 4   Time_of_Transaction               48448 non-null  float64
 5   Device_Used                       48527 non-null  object 
 6   Location                          48453 non-null  object 
 7   Previous_Fraudulent_Transactions  51000 non-null  int64  
 8   Account_Age                       51000 non-null  int64  
 9   Number_of_Transactions_Last_24H   51000 non-null  int64  
 10  Payment_Method                    48531 non-null  object 
 11  Fraudulent                        51000 non-null  int64  
dtypes: f

### What Each Column Represents<br>
Transaction_ID - arbitrary unique ID given to each transaction<br>
User_ID - The unique identifier for each user<br>
Transaction Amount - amount of money transacted<br>
Transaction Type - ATM withdrawal, Bill Payment, POS payment, bank transfer, online purchase<br>
Time of Transaction - Military time for the time the transaction occured, 0 through 23<br>
Device Used - Tablet, mobile, desktop, unknown device<br>
Location - San Francisco, New York, Chicago, Boston, Houston, Miami, Los Angeles, Seattle<br>
Previous Fraudulent Transactions - How many previous fraudelent transactions does this user have, 0 through 4<br>
Account Age - How long the account has been active in possibly months, 0 through 119<br>
Number of Transactions last 24H - The number of transactions in the last 24 hours for this user, 1 through 14<br>
Payment Method - Debit card, credit card, UPI, net banking, invalid method<br>
Fraudulent - boolean representing whether the transaction was fraudulent.


### Data cleaning

In [69]:
df["Transaction_Amount_fill_mean"] = df["Transaction_Amount"]
df["Transaction_Amount_fill_median"] = df["Transaction_Amount"]
df["Transaction_Amount_fill_mode"] = df["Transaction_Amount"]

values = {"Transaction_Amount_fill_mean": df["Transaction_Amount"].mean(), "Transaction_Amount_fill_median": df["Transaction_Amount"].median(), 
         "Transaction_Amount_fill_mode": stats.mode(df["Transaction_Amount"], nan_policy="omit").mode}
df.fillna(value = values, inplace = True)

In [ ]:
df["Transaction_Amount_fill_KNN"] = df["Transaction_Amount"]
df_drop_na = titanic_data
df_drop_na = df_drop_na.dropna(subset = "Transaction_Amount")

#x_train = df_drop_na[["SibSp", "Parch", "Fare"]]
y_train = df_drop_na["Transaction_Amount"]
#x_test = df_drop_na[["SibSp", "Parch", "Fare"]]
model = KNeighborsRegressor(n_neighbors=5)
model.fit(x_train, y_train)

answers = model.predict(x_test)
index = 0
for amount in df["Age_fill_knn"]:
    if(np.isnan(amount)):
        df.at[index, "Age_fill_knn"] = answers[index]
    index += 1

In [73]:
#Transaction_Amount, Time_of_Transaction, Device_Used, Location, Payment_Method
df.dropna(axis=0, subset=["Device_Used", "Location", "Payment_Method"], inplace = True)